### PR TITLE
Reduce memory-footprint by using alpine as base docker image

### PR DESCRIPTION
**Pull request by promehul**
Use alpine as base image.
Solves #184 #178

### CODE REVIEW

**Comment by achillesrasquinha** [comment]
This still happens to use the development server. What about the multi-staged build?

**Comment by promehul** [comment]
@achillesrasquinha I know what multi-staged build is. So what am I expected to do using this. Can u please elaborate a little.

File: Dockerfile [hunk 1]
```diff
@@ -1,40 +1,73 @@
-FROM  ubuntu:xenial
+# Use node base image
+FROM node:latest
+ENV NODE_ENV=development
 
-# Add labels for metadata
+# Create a working app directory
+RUN mkdir -p /app
```
Comment (achillesrasquinha) [inline Dockerfile]: Not required. `WORKDIR` creates this for you.

File: Dockerfile [hunk 2]
```diff
@@ -1,40 +1,73 @@
...
+WORKDIR /app
+
+# Copy the package.json file into app directory to install the npm packages
+COPY ./package.json /app
+
+# Install all the npm packages
+RUN npm install
```
Comment (achillesrasquinha) [inline Dockerfile]: `yarn` not `npm`, creates a ghost `package-lock.json` otherwise. also ensure it's `--pure-lockfile`

File: Dockerfile [hunk 3]
```diff
@@ -1,40 +1,73 @@
...
-    apt-get update &&  apt-get install java-common && apt-get install -y --no-install-recommends apt-utils && \
-    dpkg --install java-1.8.0-amazon-corretto-jdk_8.212.04-2_amd64.deb
+RUN { \
+	echo '#!/bin/sh'; \
+	echo 'set -e'; \
+	echo; \
+	echo 'dirname "$(dirname "$(readlink -f "$(which javac || which java)")")"'; \
+	} > /usr/local/bin/docker-java-home \
+	&& chmod +x /usr/local/bin/docker-java-home
+ENV JAVA_HOME /usr/lib/jvm/java-1.8-openjdk
```
Comment (achillesrasquinha) [inline Dockerfile]: Use a single layer here.

Comment (achillesrasquinha) [reply Dockerfile]:
```
ENV foo=bar \
     bar=baz
```

File: Dockerfile [hunk 4]
```diff
@@ -45,10 +78,15 @@ VOLUME /app
 # Work in the app directory of the container
 WORKDIR /app
 
+# Use docker build args to copy the static asset files built previously into this container
+RUN rm -rf /app/candis/app/assets
+COPY --from=0  /app/candis/app/assets  /app/candis/app/assets
```
Comment (achillesrasquinha) [inline Dockerfile]: create a named build, call it build.

File: Dockerfile [hunk 5]
```diff
@@ -1,40 +1,73 @@
-FROM  ubuntu:xenial
+# Use node base image
+FROM node:latest
```
Comment (achillesrasquinha) [inline Dockerfile]: name this build.

File: Dockerfile [hunk 6]
```diff
@@ -45,10 +78,15 @@ VOLUME /app
 # Work in the app directory of the container
 WORKDIR /app
 
+# Use docker build args to copy the static asset files built previously into this container
+RUN rm -rf /app/candis/app/assets
+COPY --from=0  /app/candis/app/assets  /app/candis/app/assets
+
+# Export python path
 ENV PYTHONPATH="/app/candis"
```
Comment (achillesrasquinha) [inline Dockerfile:82]: unify this.

File: package.json [hunk 7]
```diff
@@ -74,6 +74,6 @@
     "shortid": "^2.2.8",
     "socket.io": "^2.0.4",
     "store": "^2.0.12",
-    "yup": "^0.25.1"
+    "yup": "^0.27.0"
```
Comment (achillesrasquinha) [inline package.json:77]: if you're interested in updating all packages, use `yarn upgrade`

**Review by achillesrasquinha — APPROVED** [review]

File: Dockerfile (hunk 2 above)
Comment (promehul) [reply Dockerfile]: So can we use the flag ` yarn install --no-lockfile`?

File: package.json (hunk 7 above)
Comment (promehul) [reply package.json:77]: I updated all the packages using `npm-check-updates` npm package.

Comment (promehul) [reply package.json:77]: Babel is breaking on updating every package. I am trying to resolve it

File: Dockerfile [hunk 8]
```diff
@@ -1,54 +1,88 @@
...
 
-# Add labels for metadata
+# Create a working app directory
+WORKDIR /app
+
+# Copy the package.json file into app directory to install the npm packages
+COPY ./package.json /app
+
+# Install all the npm packages
+RUN yarn install && yarn upgrade
```
Comment (achillesrasquinha) [inline Dockerfile]: Why? This updates the lock file at build time.

File: Dockerfile [hunk 9]
```diff
@@ -1,54 +1,88 @@
...
+
+# Copy all the files of the project into the app directory
+COPY . /app
+
+# Run the yarn run build script
+RUN yarn build --pure-lockfile
```
Comment (achillesrasquinha) [inline Dockerfile]: [`yarn install --pure-lockfile`](https://yarnpkg.com/lang/en/docs/cli/install/#toc-yarn-install-pure-lockfile)

Comment (achillesrasquinha) [reply Dockerfile]: this should be simply `yarn build`

**Review by achillesrasquinha — CHANGES_REQUESTED** [review]

**Pull request merged**